File: frontend/src/app/pages/budget/budget.ts
import { CommonModule } from '@angular/common';
import { ChangeDetectorRef, Component, OnInit } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterLink } from '@angular/router';
import { catchError, finalize, forkJoin, Observable, of, timeout } from 'rxjs';
import { Api } from '../../services/api';
import { AppSidebarComponent } from '../../shared/app-sidebar.component';

type BudgetTab = 'dashboard' | 'planning' | 'estimates' | 'expenses' | 'monitoring' | 'report' | 'categories';

interface CostCategory {
  key: string;
  label: string;
}

@Component({
  selector: 'app-budget',
  standalone: true,
  imports: [CommonModule, FormsModule, RouterLink, AppSidebarComponent],
  templateUrl: './budget.html',
  styleUrl: './budget.css'
})
export class Budget implements OnInit {
  readonly categories: CostCategory[] = [
    { key: 'LABOR', label: 'Labor Cost' },
    { key: 'MATERIAL', label: 'Material Cost' },
    { key: 'EQUIPMENT', label: 'Equipment Cost' },
    { key: 'TRANSPORTATION', label: 'Transportation Cost' },
    { key: 'MAINTENANCE', label: 'Maintenance Cost' },
    { key: 'ADMINISTRATIVE', label: 'Administrative Cost' }
  ];

  activeTab: BudgetTab = 'dashboard';
  loading = true;
  saving = false;
  deletingId: string | null = null;
  error = '';
  message = '';
  backendReady = true;
  role = '';
  projects: any[] = [];
  selectedProjectId: number | null = null;
  budgetPlan: any = null;
  estimates: any[] = [];
  expenses: any[] = [];
  summary: any = null;

  planForm = this.emptyPlanForm();
  estimateForm = { category: 'LABOR', activity: '', estimated_amount: 0, notes: '' };
  expenseForm = { category: 'LABOR', description: '', amount: 0, expense_date: '', reference: '', source_module: 'MANUAL' };

  private loadVersion = 0;

  constructor(private api: Api, private cdr: ChangeDetectorRef) {}

  ngOnInit(): void {
    this.role = this.currentRole();
    this.loadProjects();
  }

  get canEdit(): boolean { return this.role === 'ADMIN' || this.role === 'PROJECT_MANAGER'; }
  get canView(): boolean { return ['ADMIN', 'PROJECT_MANAGER', 'SITE_ENGINEER', 'CLIENT'].includes(this.role); }
  get formTotal(): number { return this.sumPlan(); }

  loadProjects(): void {
    const version = ++this.loadVersion;
    this.loading = true;
    this.error = '';
    this.backendReady = true;

    this.api.getProjects().pipe(
      timeout(9000),
      catchError((err) => {
        this.error = this.detail(err, 'Unable to load projects. Please check whether the backend is running.');
        return of([] as any[]);
      }),
      finalize(() => this.cdr.detectChanges())
    ).subscribe((projects) => {
      if (version !== this.loadVersion) return;
      this.projects = this.list(projects);

      if (this.selectedProjectId && !this.projects.some((p) => Number(p.id) === Number(this.selectedProjectId))) {
        this.selectedProjectId = null;
      }

      if (!this.selectedProjectId && this.projects.length) {
        this.selectedProjectId = Number(this.projects[0].id);
      }

      if (this.selectedProjectId) {
        this.loadProjectFinancials();
      } else {
        this.resetFinancialData();
      }
    });
  }

  onProjectChange(): void {
    this.error = '';
    this.message = '';
    if (this.selectedProjectId) this.loadProjectFinancials();
    else this.resetFinancialData();
  }

  refresh(): void {
    if (this.selectedProjectId) this.loadProjectFinancials();
    else this.loadProjects();
  }

  loadProjectFinancials(): void {
    if (!this.selectedProjectId) {
      this.resetFinancialData();
      return;
    }

    const version = ++this.loadVersion;
    const id = Number(this.selectedProjectId);
    this.loading = true;
    this.error = '';
    this.message = '';
    this.backendReady = true;

    forkJoin({
      plan: this.safe(this.api.getBudgetPlan(id), null),
      estimates: this.safe(this.api.getCostEstimates(id), [] as any[]),
      expenses: this.safe(this.api.getExpenses(id), [] as any[]),
      summary: this.safe(this.api.getBudgetSummary(id), null)
    }).pipe(
      finalize(() => {
        if (version === this.loadVersion) this.loading = false;
        this.cdr.detectChanges();
      })
    ).subscribe((data) => {
      if (version !== this.loadVersion) return;
      this.budgetPlan = data.plan;
      this.estimates = this.list(data.estimates);
      this.expenses = this.list(data.expenses);
      this.summary = data.summary;
      this.planForm = this.toPlanForm(this.budgetPlan);
    });
  }

  saveBudgetPlan(): void {
    if (!this.canEdit || !this.selectedProjectId || this.saving) return;
    const allocations = this.categories.map((c) => ({ category: c.key, amount: this.num(this.planForm[c.key]) }));
    const total = allocations.reduce((sum, x) => sum + x.amount, 0);

    if (total <= 0) {
      this.error = 'Enter at least one category allocation.';
      return;
    }

    this.saving = true;
    this.error = '';
    this.message = '';

    const payload = {
      project_id: Number(this.selectedProjectId),
      total_budget: total,
      category_allocations: allocations
    };

    const request = this.budgetPlan?.id
      ? this.api.updateBudgetPlan(Number(this.budgetPlan.id), payload)
      : this.api.createBudgetPlan(payload);

    request.pipe(
      timeout(10000),
      finalize(() => {
        this.saving = false;
        this.cdr.detectChanges();
      })
    ).subscribe({
      next: (saved) => {
        const plan = this.normalizePlan(saved, payload);
        this.budgetPlan = plan;
        this.planForm = this.toPlanForm(plan);
        this.summary = { ...(this.summary || {}), total_budget: total, remaining_budget: total - this.amountSpent };
        this.message = 'Budget plan saved.';
        this.loadProjectFinancials();
      },
      error: (e) => {
        this.error = this.detail(e, 'Unable to save the budget plan. Please check the backend terminal for the exact error.');
      }
    });
  }

  addEstimate(): void {
    if (!this.canEdit || !this.selectedProjectId || this.saving) return;
    if (!this.estimateForm.activity.trim() || this.num(this.estimateForm.estimated_amount) <= 0) {
      this.error = 'Activity and a positive estimated amount are required.';
      return;
    }

    this.saving = true;
    this.error = '';
    this.message = '';

    const payload = {
      category: this.estimateForm.category,
      activity: this.estimateForm.activity.trim(),
      item_name: this.estimateForm.activity.trim(),
      estimated_amount: this.num(this.estimateForm.estimated_amount),
      notes: this.estimateForm.notes?.trim() || '',
      description: this.estimateForm.notes?.trim() || '',
      project_id: Number(this.selectedProjectId)
    };

    this.api.createCostEstimate(payload).pipe(
      timeout(10000),
      finalize(() => {
        this.saving = false;
        this.cdr.detectChanges();
      })
    ).subscribe({
      next: (saved) => {
        this.estimates = [...this.estimates, saved || payload];
        this.estimateForm = { category: 'LABOR', activity: '', estimated_amount: 0, notes: '' };
        this.message = 'Cost estimate added.';
        this.loadProjectFinancials();
      },
      error: (e) => {
        this.error = this.detail(e, 'Unable to save the cost estimate.');
      }
    });
  }

  deleteEstimate(item: any): void {
    if (!this.canEdit || !item?.id || !confirm('Delete this cost estimate?')) return;
    this.deletingId = `estimate-${item.id}`;
    this.api.deleteCostEstimate(Number(item.id)).pipe(
      timeout(10000),
      finalize(() => {
        this.deletingId = null;
        this.cdr.detectChanges();
      })
    ).subscribe({
      next: () => {
        this.estimates = this.estimates.filter((x) => Number(x.id) !== Number(item.id));
        this.message = 'Cost estimate deleted.';
        this.loadProjectFinancials();
      },
      error: (e) => this.error = this.detail(e, 'Unable to delete the estimate.')
    });
  }

  addExpense(): void {
    if (!this.canEdit || !this.selectedProjectId || this.saving) return;
    if (!this.expenseForm.description.trim() || this.num(this.expenseForm.amount) <= 0 || !this.expenseForm.expense_date) {
      this.error = 'Description, expense date and a positive amount are required.';
      return;
    }

    this.saving = true;
    this.error = '';
    this.message = '';

    const payload = {
      category: this.expenseForm.category,
      description: this.expenseForm.description.trim(),
      expense_name: this.expenseForm.description.trim(),
      amount: this.num(this.expenseForm.amount),
      expense_date: this.expenseForm.expense_date,
      reference: this.expenseForm.reference?.trim() || '',
      source_module: this.expenseForm.source_module || 'MANUAL',
      project_id: Number(this.selectedProjectId)
    };

    this.api.createExpense(payload).pipe(
      timeout(10000),
      finalize(() => {
        this.saving = false;
        this.cdr.detectChanges();
      })
    ).subscribe({
      next: (saved) => {
        this.expenses = [...this.expenses, saved || payload];
        this.expenseForm = { category: 'LABOR', description: '', amount: 0, expense_date: '', reference: '', source_module: 'MANUAL' };
        this.message = 'Expense recorded.';
        this.loadProjectFinancials();
      },
      error: (e) => {
        this.error = this.detail(e, 'Unable to record the expense.');
      }
    });
  }

  deleteExpense(item: any): void {
    if (!this.canEdit || !item?.id || !confirm('Delete this expense?')) return;
    this.deletingId = `expense-${item.id}`;
    this.api.deleteExpense(Number(item.id)).pipe(
      timeout(10000),
      finalize(() => {
        this.deletingId = null;
        this.cdr.detectChanges();
      })
    ).subscribe({
      next: () => {
        this.expenses = this.expenses.filter((x) => Number(x.id) !== Number(item.id));
        this.message = 'Expense deleted.';
        this.loadProjectFinancials();
      },
      error: (e) => this.error = this.detail(e, 'Unable to delete the expense.')
    });
  }

  selectTab(tab: BudgetTab): void { this.activeTab = tab; this.error = ''; this.message = ''; }

  get totalBudget(): number { return this.firstNumber(this.summary?.total_budget, this.budgetPlan?.total_budget, this.sumPlan()); }
  get estimatedCost(): number { return this.firstNumber(this.summary?.estimated_cost, this.summary?.total_estimated, this.sum(this.estimates, 'estimated_amount')); }
  get amountSpent(): number { return this.firstNumber(this.summary?.amount_spent, this.summary?.actual_cost, this.summary?.total_actual, this.sum(this.expenses, 'amount')); }
  get remainingBudget(): number { const explicit = this.numberOrNull(this.summary?.remaining_budget); return explicit ?? Math.max(0, this.totalBudget - this.amountSpent); }
  get utilization(): number { const explicit = this.numberOrNull(this.summary?.utilization_percentage); return explicit ?? (this.totalBudget > 0 ? Math.round((this.amountSpent / this.totalBudget) * 1000) / 10 : 0); }

  categoryPlanned(key: string): number {
    const row = this.findCategory(this.budgetPlan?.category_allocations || this.budgetPlan?.allocations || [], key);
    return this.firstNumber(row?.amount, row?.allocated_amount, this.planForm[key]);
  }
  categoryEstimated(key: string): number { return this.estimates.filter((x) => this.key(x.category || x.category_name) === key).reduce((s, x) => s + this.num(x.estimated_amount ?? x.amount), 0); }
  categorySpent(key: string): number { return this.expenses.filter((x) => this.key(x.category || x.category_name) === key).reduce((s, x) => s + this.num(x.amount ?? x.actual_amount), 0); }
  categoryRemaining(key: string): number { return Math.max(0, this.categoryPlanned(key) - this.categorySpent(key)); }
  categoryUtilization(key: string): number { const p = this.categoryPlanned(key); return p > 0 ? Math.min(100, Math.round((this.categorySpent(key) / p) * 1000) / 10) : 0; }

  money(value: any): string { const n = this.num(value); return new Intl.NumberFormat('en-IN', { style: 'currency', currency: 'INR', maximumFractionDigits: 0 }).format(n); }
  projectName(): string { const p = this.projects.find((x) => Number(x.id) === Number(this.selectedProjectId)); return p?.project_name || p?.name || 'Selected project'; }
  categoryLabel(value: any): string { const key = this.key(value); return this.categories.find((c) => c.key === key)?.label || String(value || 'Uncategorized'); }

  private safe<T>(request: Observable<T>, fallback: T): Observable<T> {
    return request.pipe(
      timeout(9000),
      catchError((err) => {
        this.markBackendUnavailable(err);
        return of(fallback as T);
      })
    );
  }

  private list(value: any): any[] {
    if (Array.isArray(value)) return value;
    if (Array.isArray(value?.data)) return value.data;
    if (Array.isArray(value?.items)) return value.items;
    if (Array.isArray(value?.results)) return value.results;
    return [];
  }

  private normalizePlan(saved: any, payload: any): any {
    if (saved && typeof saved === 'object') {
      return {
        ...saved,
        project_id: saved.project_id ?? payload.project_id,
        total_budget: this.num(saved.total_budget ?? payload.total_budget),
        category_allocations: Array.isArray(saved.category_allocations) ? saved.category_allocations : payload.category_allocations,
        allocations: Array.isArray(saved.allocations) ? saved.allocations : (Array.isArray(saved.category_allocations) ? saved.category_allocations : payload.category_allocations)
      };
    }
    return { id: this.budgetPlan?.id, ...payload, allocations: payload.category_allocations };
  }

  private emptyPlanForm(): any { return { LABOR: 0, MATERIAL: 0, EQUIPMENT: 0, TRANSPORTATION: 0, MAINTENANCE: 0, ADMINISTRATIVE: 0 }; }
  private toPlanForm(plan: any): any {
    const form = this.emptyPlanForm();
    const allocations = plan?.category_allocations || plan?.allocations || [];
    if (Array.isArray(allocations)) {
      allocations.forEach((row: any) => {
        const key = this.key(row.category || row.category_name || row.name);
        if (key in form) form[key] = this.num(row.amount ?? row.allocated_amount ?? row.value);
      });
    } else {
      this.categories.forEach((c) => { if (plan?.[c.key.toLowerCase()] != null) form[c.key] = this.num(plan[c.key.toLowerCase()]); });
    }
    return form;
  }
  private sumPlan(): number { return this.categories.reduce((s, c) => s + this.num(this.planForm[c.key]), 0); }
  private sum(rows: any[], field: string): number { return (rows || []).reduce((s, row) => s + this.num(row?.[field]), 0); }
  private num(v: any): number { const n = Number(v); return Number.isFinite(n) ? n : 0; }
  private numberOrNull(v: any): number | null { if (v === '' || v == null) return null; const n = Number(v); return Number.isFinite(n) ? n : null; }
  private firstNumber(...values: any[]): number { for (const v of values) { const n = this.numberOrNull(v); if (n != null) return n; } return 0; }
  private key(v: any): string { return String(v || '').trim().toUpperCase().replace(/\s+COST$/, '').replace(/\s+/g, '_'); }
  private findCategory(rows: any[], key: string): any { return (rows || []).find((x: any) => this.key(x.category || x.category_name || x.name) === key); }
  private currentRole(): string { try { return String(JSON.parse(localStorage.getItem('currentUser') || '{}')?.role || '').toUpperCase(); } catch { return ''; } }
  private markBackendUnavailable(err: any): void { if ([0, 404, 405, 501].includes(Number(err?.status))) this.backendReady = false; }
  private detail(err: any, fallback: string): string {
    const detail = err?.error?.detail ?? err?.error?.message ?? err?.message;
    if (Array.isArray(detail)) return detail.map((x: any) => x?.msg || JSON.stringify(x)).join(', ');
    return detail || fallback;
  }
  private resetFinancialData(): void {
    this.budgetPlan = null;
    this.estimates = [];
    this.expenses = [];
    this.summary = null;
    this.planForm = this.emptyPlanForm();
    this.loading = false;
    this.cdr.detectChanges();
  }
}

File: backend/app/api/budget.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func
from sqlalchemy.orm import Session
from app.database.database import get_db
from app.models.project import Project
from app.models.budget import Budget
from app.models.budget_category import BudgetCategory
from app.models.cost_estimate import CostEstimate
from app.models.expense import Expense
from app.schemas.budget import BudgetPlanCreate, BudgetPlanUpdate
from app.core.permissions import role_required

router=APIRouter(prefix="/budgets",tags=["Budgets"])
READ=role_required(["ADMIN","PROJECT_MANAGER","SITE_ENGINEER","CLIENT"])
EDIT=role_required(["ADMIN","PROJECT_MANAGER"])
CANONICAL=[("LABOR","Labor Cost"),("MATERIAL","Material Cost"),("EQUIPMENT","Equipment Cost"),("TRANSPORTATION","Transportation Cost"),("MAINTENANCE","Maintenance Cost"),("ADMINISTRATIVE","Administrative Cost")]

def _key(v): return str(v or '').strip().upper().replace(' COST','').replace(' ','_')
def _category(db,name):
    key=_key(name)
    label=next((label for k,label in CANONICAL if k==key), str(name or key).replace('_',' ').title())
    row=db.query(BudgetCategory).filter(func.upper(BudgetCategory.name)==label.upper()).first()
    if not row:
        row=BudgetCategory(name=label); db.add(row); db.flush()
    return row

def _project(db,pid):
    p=db.query(Project).filter(Project.id==pid).first()
    if not p: raise HTTPException(404,"Project not found")
    return p

def _serialize_plan(db,pid):
    rows=db.query(Budget).filter(Budget.project_id==pid).order_by(Budget.id).all()
    if not rows: return None
    alloc=[]
    for r in rows:
        name=r.category.name if r.category else "Uncategorized"
        alloc.append({"category":_key(name),"category_name":name,"category_id":r.category_id,"amount":float(r.allocated_amount or 0),"allocated_amount":float(r.allocated_amount or 0)})
    return {"id":rows[0].id,"project_id":pid,"total_budget":round(sum(x["amount"] for x in alloc),2),"category_allocations":alloc,"allocations":alloc,"status":rows[0].status or "Active","description":rows[0].description,"start_date":rows[0].start_date,"end_date":rows[0].end_date}

def _save_plan(db,pid,data):
    _project(db,pid)
    payload=data.model_dump(exclude_unset=True)
    allocations=payload.get('category_allocations') or []
    if not allocations and payload.get('allocated_amount') is not None:
        allocations=[{"category":"ADMINISTRATIVE","amount":payload['allocated_amount']}]
    if not allocations and payload.get('total_budget') is not None:
        allocations=[{"category":"ADMINISTRATIVE","amount":payload['total_budget']}]
    if not allocations:
        raise HTTPException(422,"At least one category allocation is required")

    cleaned=[]
    for item in allocations:
        item=item if isinstance(item,dict) else item.model_dump()
        category=item.get('category') or item.get('category_name') or item.get('name')
        amount=float(item.get('amount') if item.get('amount') is not None else item.get('allocated_amount') or 0)
        if amount < 0:
            raise HTTPException(422,"Budget allocation amount cannot be negative")
        cleaned.append({"category":category,"amount":amount})

    if sum(x["amount"] for x in cleaned) <= 0:
        raise HTTPException(422,"Total budget must be greater than zero")

    db.query(Budget).filter(Budget.project_id==pid).delete(synchronize_session=False)
    db.flush()
    for item in cleaned:
        cat=_category(db,item.get('category'))
        db.add(Budget(
            project_id=pid,
            category_id=cat.id,
            allocated_amount=item["amount"],
            description=payload.get('description'),
            status=payload.get('status') or 'Active',
            start_date=payload.get('start_date'),
            end_date=payload.get('end_date')
        ))
    db.commit()
    return _serialize_plan(db,pid)

@router.get("/project/{project_id}")
def get_budget_plan(project_id:int,db:Session=Depends(get_db),_=Depends(READ)):
    _project(db,project_id); return _serialize_plan(db,project_id)

@router.get("/project/{project_id}/summary")
def get_budget_summary(project_id:int,db:Session=Depends(get_db),_=Depends(READ)):
    _project(db,project_id)
    total=float(db.query(func.coalesce(func.sum(Budget.allocated_amount),0)).filter(Budget.project_id==project_id).scalar() or 0)
    estimated=float(db.query(func.coalesce(func.sum(CostEstimate.estimated_amount),0)).filter(CostEstimate.project_id==project_id).scalar() or 0)
    spent=float(db.query(func.coalesce(func.sum(Expense.amount),0)).filter(Expense.project_id==project_id).scalar() or 0)
    util=round(spent/total*100,2) if total else 0
    status='Over Budget' if spent>total and total else ('Critical' if util>=90 else 'Warning' if util>=75 else 'Within Budget')
    return {"project_id":project_id,"total_budget":round(total,2),"estimated_cost":round(estimated,2),"total_estimated":round(estimated,2),"amount_spent":round(spent,2),"actual_cost":round(spent,2),"total_actual":round(spent,2),"remaining_budget":round(total-spent,2),"utilization_percentage":util,"status":status}

@router.post("/")
def create_budget(data:BudgetPlanCreate,db:Session=Depends(get_db),_=Depends(EDIT)):
    # Upsert behavior keeps the Angular page simple and prevents a stale/null
    # budgetPlan from turning a second save into a 409 conflict.
    return _save_plan(db,data.project_id,data)

@router.get("/")
def get_budgets(project_id:int|None=None,db:Session=Depends(get_db),_=Depends(READ)):
    if project_id is not None: return [_serialize_plan(db,project_id)] if _serialize_plan(db,project_id) else []
    ids=[x[0] for x in db.query(Budget.project_id).distinct().all()]; return [_serialize_plan(db,pid) for pid in ids]

@router.get("/{budget_id}")
def get_budget_by_id(budget_id:int,db:Session=Depends(get_db),_=Depends(READ)):
    """Return the complete budget plan that contains the requested budget row."""
    row=db.query(Budget).filter(Budget.id==budget_id).first()
    if not row: raise HTTPException(404,"Budget plan not found")
    return _serialize_plan(db,row.project_id)

@router.put("/{budget_id}")
def update_budget(budget_id:int,data:BudgetPlanUpdate,db:Session=Depends(get_db),_=Depends(EDIT)):
    row=db.query(Budget).filter(Budget.id==budget_id).first()
    if not row: raise HTTPException(404,"Budget plan not found")
    return _save_plan(db,row.project_id,data)

@router.delete("/{budget_id}")
def delete_budget(budget_id:int,db:Session=Depends(get_db),_=Depends(EDIT)):
    row=db.query(Budget).filter(Budget.id==budget_id).first()
    if not row: raise HTTPException(404,"Budget plan not found")
    pid=row.project_id; db.query(Budget).filter(Budget.project_id==pid).delete(synchronize_session=False); db.commit(); return {"message":"Budget plan deleted successfully"}
